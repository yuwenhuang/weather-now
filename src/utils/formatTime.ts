export const formatTime = (s: number): string => {
    const dtFormat = new Intl.DateTimeFormat('en-GB', {
      timeStyle: 'medium',
      timeZone: 'UTC'
    });
    const result = dtFormat.format(new Date(s * 1e3))
  
    return result
}