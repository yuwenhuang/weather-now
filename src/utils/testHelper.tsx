import React from 'react'
import { ThemeProvider } from 'styled-components'
import { MemoryRouter } from "react-router-dom"

import { weatherTheme } from '../styles/weatherTheme'

export const wrapper = (child: React.Component) => {
  return( 
  <ThemeProvider theme={weatherTheme}><MemoryRouter>{child}</MemoryRouter></ThemeProvider>
  )
}
