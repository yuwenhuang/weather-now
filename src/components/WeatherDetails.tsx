import React, { useState, useEffect }  from 'react';
import {
  useParams,
  useLocation,
} from 'react-router-dom'
import { ICityWeather } from '../types'
import { formatTime } from '../utils/formatTime'

interface RouteParams {
  name?: string;
}
interface LocationParams {
  cityWeather: ICityWeather
}

const WeatherDetails =  () => {
  const { name } = useParams<RouteParams>()
  const location = useLocation<LocationParams>()
  const { cityWeather } = location.state
  const { main, visibility, sys, timezone } = cityWeather
  const { sunrise, sunset } = sys
  const localSunriseTime = formatTime(sunrise + timezone)
  const localSunsetTime = formatTime(sunset + timezone)

  return (
    <>
      <p>Weather Details: {name}</p>
      <p>temperature: {main.temp}</p>
      <p>humidity: {main.humidity} %</p>
      <p>visibility: {visibility} m</p>
      <p>sunrise: {localSunriseTime}</p>
      <p>sunset: {localSunsetTime}</p>
    </>
  ) 
  }


export default WeatherDetails;
