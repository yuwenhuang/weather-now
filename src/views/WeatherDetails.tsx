import React from 'react';
import {
  useParams,
  useLocation,
} from 'react-router-dom'
import styled from 'styled-components'

import { CityWeather } from '../types'
import { formatTime } from '../utils/formatTime'
import { media } from '../styles/mediaQuery'

import { WeatherSummary } from '../components/WeatherSummary'
import { WeatherCondition } from '../components/WeatherCondition'

const WeatherDetailsWrapper = styled.section`
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: 20px;
  padding: 16px;
`;

const WeatherDetailsTitle = styled.h1`
  font-size: 3.5em;
  text-align: center;
  color: ${(props) => props.theme.colors.blackCoffee};
`
const WeatherSetion = styled.div`
  display: flex;
  ${media.desktop} {
    flex-direction: row;
  }
  ${media.tablet} {
    flex-direction: row;
  }
  ${media.phone} {
    flex-direction: column;
  }
`;

const WeatherConditions = styled.div`
  display: flex;
  flex-direction: column;
`
const WeatherConditionsRow = styled.div`
  display: flex;
  ${media.desktop} {
    flex-direction: row;
  }
  ${media.tablet} {
    flex-direction: row;
  }
  ${media.phone} {
    flex-direction: column;
  }
`

interface RouteParams {
  name?: string;
}
interface LocationParams {
  cityWeather: CityWeather
}

const WeatherDetails =  (): JSX.Element  => {
  const { name } = useParams<RouteParams>()
  const location = useLocation<LocationParams>()
  const { cityWeather } = location.state
  const { main, visibility, sys, timezone, weather } = cityWeather
  const { temp, temp_min, temp_max } = main
  const { sunrise, sunset } = sys
  const { description } = weather[0]
  const localSunriseTime = formatTime(sunrise + timezone)
  const localSunsetTime = formatTime(sunset + timezone)

  return (
    <WeatherDetailsWrapper>
      <WeatherDetailsTitle>{name}</WeatherDetailsTitle>
      <WeatherSetion>
        <WeatherSummary temp={temp} temp_min={temp_min} temp_max={temp_max} description={description}/>
        <WeatherConditions>
          <WeatherConditionsRow>
            <WeatherCondition description='humidity' value={main.humidity} unit='%'/>
            <WeatherCondition description='visibility' value={visibility} unit='m'/>
          </WeatherConditionsRow>
          <WeatherConditionsRow>
            <WeatherCondition description='sunrise' value={localSunriseTime} />
            <WeatherCondition description='sunset' value={localSunsetTime} />  
          </WeatherConditionsRow>
        </WeatherConditions>
      </WeatherSetion>    
    </WeatherDetailsWrapper>
  ) 
}


export default WeatherDetails;
