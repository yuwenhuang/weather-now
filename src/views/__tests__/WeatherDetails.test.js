import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components'

import { wrapper } from '../../utils/testHelper'
import WeatherDetails from '../WeatherDetails'

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: () => ({
    state: {
      cityWeather: {
        name: 'test city',
        main: {
          temp: '10',
          temp_min: '0',
          temp_max: '20',
          humidity: '25'
        },
        visibility: '1000',
        sys: {
          sunrise: 1609459200,
          sunset: 1609459200,
          country: 'test country'
        },
        weather: [{
          id: 1,
          main: 'main',
          description: 'test',
          icon: 'test'
        }],
        timezone: 1609459200
      }
    }  
  })
}));

it('renders <WeatherDetails /> component', () => {
  const component = renderer.create(
    wrapper(<WeatherDetails />)
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
})