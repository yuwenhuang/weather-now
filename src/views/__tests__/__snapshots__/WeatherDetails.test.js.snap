// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders <WeatherDetails /> component 1`] = `
.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-aligh-items: center;
  -webkit-box-aligh: center;
  -ms-flex-aligh: center;
  aligh-items: center;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
  background-color: #99E1D9;
  border-radius: 10px;
  box-shadow: 5px 5px 15px 5px;
  margin: 20px;
  padding: 16px;
}

.c4 {
  text-align: center;
  font-size: 1em;
}

.c5 {
  text-align: center;
  font-size: 3em;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c7 {
  padding: 16px;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  background-color: #99E1D9;
  border-radius: 10px;
  min-width: 150px;
  margin: 20px;
  padding: 16px;
  box-shadow: 5px 5px 15px 5px;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 10px;
}

.c12 {
  font-size: 1.5em;
  text-align: center;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 20px;
  padding: 16px;
}

.c1 {
  font-size: 3.5em;
  text-align: center;
  color: #32292F;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media (max-width:2560px) {
  .c2 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (max-width:768px) {
  .c2 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (max-width:576px) {
  .c2 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

@media (max-width:2560px) {
  .c9 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (max-width:768px) {
  .c9 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
  }
}

@media (max-width:576px) {
  .c9 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

<div
  className="c0"
>
  <h1
    className="c1"
  />
  <div
    className="c2"
  >
    <div
      className="c3"
    >
      <text
        className="c4"
      >
        test
      </text>
      <text
        className="c5"
      >
        10
        °C
      </text>
      <div
        className="c6"
      >
        <div
          className="c7"
        >
          H:
          20
          °C
        </div>
        <div
          className="c7"
        >
          L:
          0
          °C
        </div>
      </div>
    </div>
    <div
      className="c8"
    >
      <div
        className="c9"
      >
        <div
          className="c10"
        >
          <div
            className="c11"
          >
            humidity
          </div>
          <text
            className="c12"
          >
            25
             
            %
          </text>
        </div>
        <div
          className="c10"
        >
          <div
            className="c11"
          >
            visibility
          </div>
          <text
            className="c12"
          >
            1000
             
            m
          </text>
        </div>
      </div>
      <div
        className="c9"
      >
        <div
          className="c10"
        >
          <div
            className="c11"
          >
            sunrise
          </div>
          <text
            className="c12"
          >
            12:00 AM
             
          </text>
        </div>
        <div
          className="c10"
        >
          <div
            className="c11"
          >
            sunset
          </div>
          <text
            className="c12"
          >
            12:00 AM
             
          </text>
        </div>
      </div>
    </div>
  </div>
</div>
`;
